### PR TITLE
Allow migrating on a clickhouse cluster

### PR DESCRIPTION
**Pull request by macobo**
Context under https://github.com/PostHog/posthog/issues/4090

Do not merge this yet - the secrets do not yet contain the right values, depends on answer to https://github.com/PostHog/internal/issues/268

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

**Comment by macobo** [comment]
I've updated the production secret to include the hosts (3-8) based on the linked issue + added some extra code for extra safety. This is ready to be merged and get rolled out :crossed_fingers:

File: posthog/settings.py [hunk 1]
```diff
@@ -157,7 +157,12 @@ def print_warning(warning_lines: Sequence[str]):
     _clickhouse_http_protocol = "https://"
     _clickhouse_http_port = "8443"
 
-CLICKHOUSE_HTTP_URL = _clickhouse_http_protocol + CLICKHOUSE_HOST + ":" + _clickhouse_http_port + "/"
+CLICKHOUSE_HTTP_URL = f"{_clickhouse_http_protocol}{CLICKHOUSE_HOST}:{_clickhouse_http_port}/"
+
+_clickhouse_hosts = get_from_env("CLICKHOUSE_MIGRATION_HOSTS", CLICKHOUSE_HOST).split(",")
+CLICKHOUSE_MIGRATION_HOSTS_HTTP_URLS = [
+    f"{_clickhouse_http_protocol}{host}:{_clickhouse_http_port}/" for host in _clickhouse_hosts
+]
```
Comment (tiina303) [inline posthog/settings.py:165]: Maybe consider making `f"{_clickhouse_http_protocol}{host}:{_clickhouse_http_port}/"` a function with `host` param, so it's easier to read/edit later?

File: task-definition.migration.json [hunk 2]
```diff
@@ -106,6 +106,10 @@
                     "name": "CLICKHOUSE_PASSWORD",
                     "valueFrom": "arn:aws:secretsmanager:us-east-1:795637471508:secret:Posthog_Production_Heroku-FQ2itU:CLICKHOUSE_PASSWORD::"
                 },
+                {
+                    "name": "CLICKHOUSE_MIGRATION_HOSTS",
+                    "valueFrom": "arn:aws:secretsmanager:us-east-1:795637471508:secret:Posthog_Production_Heroku-FQ2itU:CLICKHOUSE_MIGRATION_HOSTS::"
```
Comment (tiina303) [inline task-definition.migration.json:111]: Where do these come from?

File: ee/management/commands/migrate_clickhouse.py [hunk 3]
```diff
@@ -21,21 +27,36 @@ def add_arguments(self, parser):
...
-            for migration_name in self.get_migrations(database, options["upto"]):
+            migrations = list(self.get_migrations(database, options["upto"]))
+            for migration_name, operations in migrations:
                 print(f"Migration would get applied: {migration_name}")
-            else:
+                for op in operations:
+                    sql = getattr(op, "_sql")
+                    if options["print_sql"] and sql is not None:
+                        print(indent("\n\n".join(sql), "    "))
+            if len(migrations) == 0:
```
Comment (tiina303) [inline ee/management/commands/migrate_clickhouse.py:59]: why did you change the `else` to `if len(migrations) == 0`?

**Review by tiina303 — APPROVED** [review]
Generally looks fine to me (didn't fully check the sql output part, but since that was under `--plan` it can't cause that much harm.
Did you test this somehow?

File: task-definition.migration.json (hunk 2 above)
Comment (macobo) [reply task-definition.migration.json:111]: In general this section sets environment variables set in the ECS container running posthog production.

The value in `valueFrom` is a AWS secrets manager ARN which aws will resolve when creating the task. We keep our secrets in there.

In this case this is a list of hosts we host clickhouse on since we need to do the schema changes on all of them.

File: ee/management/commands/migrate_clickhouse.py (hunk 3 above)
Comment (macobo) [reply ee/management/commands/migrate_clickhouse.py:59]: Previous version was slightly buggy - `for` loop `else` fires always after end (unless you break) which was not what I intended. https://book.pythontips.com/en/latest/for_-_else.html

**Comment by macobo** [comment]
Yep, the changes were tested locally except for the added secret references.

File: posthog/settings.py (hunk 1 above)
Comment (macobo) [reply posthog/settings.py:165]: In general I agree, but I'd avoid introducing functions in settings.py due to convention. Given there are only two instances of this + they're close by I don't think it's that bad yet. :)

Comment (tiina303) [reply posthog/settings.py:165]: totally fine to leave it as is.

n00b q: what convention is that as I do see other functions in this file?
